stl/inc/chrono

```diff
@@ -619,27 +671,1268 @@ namespace chrono {
+        // doing everything in for > using a while loop
+        for (; _Num % _Den != 0 && _Result < 19; _Num = _Num % _Den * 10, ++_Result) {
+        }
+        return _Result == 19 ? 6 : _Result;
+    }
+
+    constexpr intmax_t _Pow10(const int _Exp) {
+        intmax_t _Result = 1;
+        for (int _Ix = 0; _Ix < _Exp; ++_Ix) {
+            _Result *= 10;
+        }
+        return _Result;
+    }
+
+    template <class _Duration>
+    requires _Is_duration_v<_Duration> class hh_mm_ss {
+    public:
+        static constexpr unsigned fractional_width = _Fractional_width(_Duration::period::num, _Duration::period::den);
+        using precision =
+            duration<common_type_t<typename _Duration::rep, seconds::rep>, ratio<1, _Pow10(fractional_width)>>;
+
+        constexpr hh_mm_ss() noexcept : hh_mm_ss{_Duration::zero()} {}
+        // clang-format off
+        constexpr explicit hh_mm_ss(_Duration _Dur)
+            : _Is_neg{_Dur < _Duration::zero()},
+              _Hours{_CHRONO duration_cast<_CHRONO hours>(_CHRONO abs(_Dur))},
+              _Mins{_CHRONO duration_cast<_CHRONO minutes>(_CHRONO abs(_Dur) - hours())},
+              _Secs{_CHRONO duration_cast<_CHRONO seconds>(_CHRONO abs(_Dur) - hours() - minutes())} {
+            // clang-format on
+            if constexpr (treat_as_floating_point_v<typename precision::rep>) {
+                _Sub_secs = _CHRONO abs(_Dur) - hours() - minutes() - seconds();
+            } else {
+                _Sub_secs = _CHRONO duration_cast<precision>(_CHRONO abs(_Dur) - hours() - minutes() - seconds());
+            }
+        }
+
+        _NODISCARD constexpr bool is_negative() const noexcept {
+            return _Is_neg;
+        }
+        _NODISCARD constexpr hours hours() const noexcept {
+            return _Hours;
+        }
+        _NODISCARD constexpr minutes minutes() const noexcept {
+            return _Mins;
+        }
+        _NODISCARD constexpr seconds seconds() const noexcept {
+            return _Secs;
+        }
+        _NODISCARD constexpr precision subseconds() const noexcept {
+            return _Sub_secs;
+        }
+
+        _NODISCARD constexpr explicit operator precision() const noexcept {
+            return to_duration();
+        }
+        _NODISCARD constexpr precision to_duration() const noexcept {
+            const auto _Dur = _Hours + _Mins + _Secs + _Sub_secs;
+            return _Is_neg ? -_Dur : _Dur;
+        }
+
+    private:
+        bool _Is_neg;
+        _CHRONO hours _Hours;
+        _CHRONO minutes _Mins;
+        _CHRONO seconds _Secs;
+        precision _Sub_secs;
+    };
+
+    _NODISCARD constexpr bool is_am(const hours& _Hours) noexcept {
+        return _Hours >= hours{0} && _Hours <= hours{11};
+    }
+    _NODISCARD constexpr bool is_pm(const hours& _Hours) noexcept {
+        return _Hours >= hours{12} && _Hours <= hours{23};
+    }
+
+    _NODISCARD constexpr hours make12(const hours& _Hours) noexcept {

```

Comment:
Sorry I'm late to the party, but I was working on some `<chrono>` items myself and found [time.12] features already here. FWIW, my implementations are below, which in particular avoid mod in `make12`. Codegen compared at https://godbolt.org/z/43P6de. It can also be done with `hours` objects instead of `hours::rep`, though that seems to result in fewer `cmov`s and spilling the return value from a register.
```cpp
    _NODISCARD hours make12_alt(const hours& _H) noexcept {
        const auto _H_count{_H.count()};
        auto _Ret{_H_count == 0 ? 12 : _H_count};
        if (_Ret > 12)
            _Ret -= 12;
        return hours{_Ret};
    }

    _NODISCARD hours make24_alt(const hours& _H, bool _Is_pm) noexcept {
        const auto _H_count{_H.count()};
        auto _Ret{_H_count == 12 ? 0 : _H_count};
        if (_Is_pm)
            _Ret += 12;
        return hours{_Ret};
    }
```

Reply:
That does indeed look better than my naive approach, thanks.